# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/tests/browser/windows/ia2/browser_textSelectionContainer.js
# Commit: 547c6d91f7e9
# Full Hash: 547c6d91f7e99909a4ec3f96327251641705d217
# Author: James Teh <jteh@mozilla.com>
# Date: 2025-05-07 04:19:47
# Regressor Bug: 1869966
# File Overlap Count: 1
# Description:
#   Bug 1869966 part 3: When calculating a DOM point from a TextLeafPoint for a non-text node, use the child index within the parent, not an offset within the node itself. r=eeejay
#   
#   For text nodes, DOM points use the character offset within the node itself, just like accessibility does.
#   For non-text nodes, DOM points use the child index within the parent to specify the position before or after the node.
#   Previously, TextLeafPoint::ToDOMPoint always used an offset within the node itself.
# ==============================================================================

diff -r 16652a9ee9f7 -r 547c6d91f7e9 accessible/tests/browser/windows/ia2/browser_textSelectionContainer.js
--- a/accessible/tests/browser/windows/ia2/browser_textSelectionContainer.js	Tue May 06 22:13:58 2025 +0000
+++ b/accessible/tests/browser/windows/ia2/browser_textSelectionContainer.js	Tue May 06 22:13:58 2025 +0000
@@ -21,7 +21,7 @@
  * Test IAccessibleTextSelectionContainer::setSelections.
  */
 addAccessibleTask(
-  `<p id="p">ab<a id="link" href="/">cd</a>ef</p>`,
+  `<p id="p">ab<a id="link" href="/">cd</a>ef<img id="img" src="https://example.com/a11y/accessible/tests/mochitest/moz.png" alt="g"></p>`,
   async function testSetSelections(browser, docAcc) {
     docAcc.QueryInterface(nsIAccessibleText);
     await runPython(`
@@ -66,6 +66,21 @@
     await selected;
     checkSelection(docAcc, [[link, 1, p, 4]]);
 
+    info("Selecting f");
+    selected = waitForEvent(EVENT_TEXT_SELECTION_CHANGED, p);
+    await runPython(`
+      docSel.setSelections(1, byref(IA2TextSelection(p, 4, p, 5, False)))
+    `);
+    await selected;
+    checkSelection(docAcc, [[p, 4, p, 5]]);
+    // DOM treats an end point of (img, 0) as including the image. Ensure we
+    // used a DOM child index.
+    await invokeContentTask(browser, [], () => {
+      const sel = content.getSelection();
+      is(sel.focusNode.id, "p", "DOM selection focus node correct");
+      is(sel.focusOffset, 3, "DOM selection focus offset correct");
+    });
+
     info("Selecting a, c");
     selected = waitForEvent(EVENT_TEXT_SELECTION_CHANGED, link);
     await runPython(`
